ledger/README.md

```diff
@@ -7,3 +7,58 @@ This package provides two ledger implementations:
 
 - **Partial Ledger** implements the ledger functionality for a limited subset of keys. Partial ledgers are designed to be constructed and verified by a collection of proofs from a complete ledger. The partial ledger uses a partial binary Merkle trie which holds intermediate hash value for the pruned branched and prevents updates to keys that were not part of proofs.
+
+## Definitions
+
+### binary Merkle tree
+
+![binary Merkle tree image](/ledger/docs/binary_merkle_tree.png?raw=true "binary Merkle tree" )
+
+In this context a *binary Merkle tree* is defined as a full binary tree with an specific height including three type of nodes:
+
+- leaf nodes: holds a payload (data), a path (where the node is located), and a hash value (hash of path and payload content)
+
+- empty leaf nodes: doesn't hold any data and only stores a path, and a default hash value based on the height of tree
+
+- intermediate nodes: holds a path and a hash value which is defined as hash of hash value of left and right children.
+
+![node types image](/ledger/docs/node_types.png)
+
+A *path* is a unique address of a node storing a payload. Paths are derived from the content of payloads (see common/pathfinder).
+
+![paths image](/ledger/docs/paths.png?raw=true "paths")
+
+#### Operations
+
+**Get**: Fetching a payload from the binary Merkle tree is by traversing the tree based on path bits. (0: left branch, 1: right branch)
+
+**Update**: Updates to the tree starts with traversing the tree to the leaf node, updating payload, hash value of that node and hash value of all the ancestor nodes (nodes on higher level connected to this node).
+
+![update image](/ledger/docs/tree_update.gif?raw=true "update")
+
+**Prove**: A binary Merkle tree can provide an inclusion proof for any given payload. A Merkle proof in this context includes all the information needed to walk through a tree branch from an specific leaf node (key) up to the root of the tree (yellow node hash values are needed for inclusion proof for the green node).
+
+![proof image](/ledger/docs/proof.png?raw=true "proof")
+
+### binary Merkle trie
+A **binary Merkle trie** in this context is defined as a compact version of binary Merkle tree, providing exact same functionality but don’t store unnecessary nodes.
+
+![binary partial trie image](/ledger/docs/trie_update.gif?raw=true "binary partial trie")
+
+### forest 
+A **forest** holds a set of binary Merkle tries, given a parent trie and a batch of updates it creates a new binary merkle trie and adds it to the forest.
```

Comment:
This is how we _use_ the forrest. However, a forest is just a specific type of graph. There is no concept of updates contained in the graph-theoretical definition:
> [[source](https://mathworld.wolfram.com/Forest.html)] A forest is an acyclic graph. Forests therefore consist only of (possibly disconnected) trees, hence the name "forest." 

```suggestion
Formally, a **forest** is any acyclic graph. Any set of disjoint trees forms a forest. In the context of Flow, we take an existing state, represented by a Merkle tree. Updating the payload of some of the leafs creates a new Merkle tree, which we add to the forest. In other words, the forest holds a set of state snapshots
```

Reply:
updated at 0d1658b